api/rpc/stack/stack-state.go

```diff
@@ -0,0 +1,98 @@
+package stack
+
+import (
+	"fmt"
+	"github.com/appcelerator/amp/api/runtime"
+	"golang.org/x/net/context"
+)
+
+const (
+	// Stopped (initial state)
+	Stopped = iota
+	// Starting state
+	Starting
+	// Running  state
+	Running
+	// Redeploying  state
+	Redeploying
+)
+
+var stackRuleSet = RuleSet{
+	//   | Stopped   | Starting  | Running   | Redeploying
+	[]bool{false /**/, true /* */, false /**/, true /* */}, // Stopped (initial state)
+	[]bool{false /**/, false /**/, true /* */, false /**/}, // Starting
+	[]bool{true /* */, false /**/, false /**/, true /* */}, // Running
+	[]bool{true /* */, true /* */, false /**/, false /**/}, // Redeploying
+}
+
+var stackStateMachine = NewStateMachine(stackRuleSet)
+
+// IsStopped returns true if the state is Stopped
+func (s *Stack) IsStopped() (bool, error) { return s.testState(Stopped) }
+
+// IsStarting returns true if the state is Starting
+func (s *Stack) IsStarting() (bool, error) { return s.testState(Starting) }
+
+// IsRunning returns true if the state is Running
+func (s *Stack) IsRunning() (bool, error) { return s.testState(Running) }
+
+// IsRedeploying returns true if the state is Redeploying
+func (s *Stack) IsRedeploying() (bool, error) { return s.testState(Redeploying) }
+
+// SetStopped sets the state to "stopped".
+func (s *Stack) SetStopped() error { return s.transitionTo(Stopped) }
+
+// SetStarting sets the state to "starting".
+func (s *Stack) SetStarting() error { return s.transitionTo(Starting) }
+
+// SetRunning sets the state to "running".
+func (s *Stack) SetRunning() error { return s.transitionTo(Running) }
+
+// SetRedeploying sets the state to "redeploying".
+func (s *Stack) SetRedeploying() error { return s.transitionTo(Redeploying) }
+
+func (s *Stack) getState() (int32, error) {
+	stack := &Stack{}
+	err := runtime.Store.Get(context.Background(), "stacks/"+s.Id, stack, false)
+	if err != nil {
+		return -1, err
+	}
+	return stack.State, nil
+}
+
+func (s *Stack) setState(state int32) error {
+	stack := &Stack{}
```

Comment:
Technically we will need to do get/set atomically (part of a transaction). You don't want the state changing during the processing of this request. We can add it later, but please leave a TODO comment. (This is a general comment that applies to `transitionTo`, etc).